broker-core/src/main/java/io/zeebe/broker/job/processor/JobInstanceStreamProcessor.java

```diff
@@ -65,16 +77,99 @@ public TypedStreamProcessor createStreamProcessor(TypedStreamEnvironment environ
         .onCommand(ValueType.JOB, JobIntent.TIME_OUT, new TimeOutJobProcessor())
         .onCommand(ValueType.JOB, JobIntent.UPDATE_RETRIES, new UpdateRetriesJobProcessor())
         .onCommand(ValueType.JOB, JobIntent.CANCEL, new CancelJobProcessor())
-        .withStateResource(jobIndex.getMap())
+        .withRocksDBController(dbController)
+        .withListener(this)
         .build();
   }
 
+  public StateSnapshotController createSnapshotController(StateStorage storage) {
+    return new StateSnapshotController(dbController, storage);
+  }
+
+  @Override
+  public void onOpen(TypedStreamProcessor streamProcessor) {
+    if (!dbController.isOpened()) {
+      throw new IllegalStateException("should have opened internal state database");
+    }
+
+    db = dbController.getDb();
+    recoverLatestJobKey(streamProcessor.getKeyGenerator());
+  }
+
+  @Override
+  public void onClose() {
+    dbController.close();
+    db = null;
+  }
+
+  private void recoverLatestJobKey(KeyGenerator keyGenerator) {
```

Comment:
It is, but the key generator is not itself persisted currently if you use RocksDB (it needs to be persisted in the snapshot since you need to keep track of the `nextKey`). Hence why I keep it in RocksDB. It's not ideal at the moment (it's a little ugly in fact), but I think we can live with it until we migrate stream processors to use RocksDB, at which point we can revise whether we need a KeyGenerator or just keep track of it in RocksDB using the merge operator (read + update in one operation).